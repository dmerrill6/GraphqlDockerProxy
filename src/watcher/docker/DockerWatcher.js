//@flow

import type { Endpoints } from '../../finder/findEndpointsInterface'
import cloner from 'cloner'
import * as clientLabels from '../../finder/clientLabels'
import { token, network } from '../../properties'

import monitor from 'node-docker-monitor'
declare function dataUpdatedListener(data: Endpoints): void
declare function idx(obj: any, callBack: any):any

export class DockerWatcher{
  endpoints : Endpoints = {}
  dataUpdatedListener: dataUpdatedListener = () => {}

  constructor(){}
  __callDataUpdateListener = async() => {
    const realEndpoint = cloner.deep.copy(this.endpoints)
    for (const one in realEndpoint){
      if (realEndpoint[one].length == 0){
        delete realEndpoint[one]
      }

    }
    console.log(realEndpoint)
    this.dataUpdatedListener(realEndpoint)

  }


  /**
 * Schaut ob es sich um eine absolute url handelt.(Startet mit http(s)://)
 * Wenn nicht sucht sie die ip des netzwerkes raus. (Relative URL z.B.: :{port}{suburl}(:3001/graphql))
 *
 *
 */
 updateUrl = (url:string, sockData:any) :string => {
   if (url.startsWith('http')){
     return url
   } else {
     return 'http://' + sockData.NetworkSettings.Networks[network()].IPAddress + url
   }
 }

 __deleteEndpoint = (namespace: string, deploymentName: string) => {
   if (this.endpoints[namespace] == undefined){
     return
   }
   for (let i = 0 ; i < this.endpoints[namespace].length; i++){
     if (this.endpoints[namespace][i].__deploymentName == deploymentName){
       this.endpoints[namespace].splice(i, 1)
     }
   }
   if (this.endpoints[namespace].length === 0){
     delete this.endpoints[namespace]
   }
 }

  __onContainerUp = (container: any) => {
    if (container.Labels[clientLabels.TOKEN] == token()){
      const url = container.Labels[clientLabels.URL]
      const namespace = container.Labels[clientLabels.NAMESPACE]
      const deploymentName = container.Id
      this.__deleteEndpoint(namespace, deploymentName)

      if (this.endpoints[namespace] == undefined){
        this.endpoints[namespace] = []
      }
      // $FlowFixMe: suppressing this error until we can refactor
      this.endpoints[namespace].push({
        url: this.updateUrl(url, container),
        namespace,
        typePrefix: namespace + '_',
        __created: container.Created,
        __imageID: container.Image,
        __deploymentName: deploymentName,
      })
      this.__callDataUpdateListener()
    }
  }

  __onContainerDown = (container) => {}
  watchEndpoint = () => {
    monitor({
      onContainerUp: this.__onContainerUp,

      onContainerDown: this.__onContainerDown,
    })
  }
}
