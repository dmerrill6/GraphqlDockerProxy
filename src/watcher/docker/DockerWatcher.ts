
import * as clientLabels from '../../finder/clientLabels';
import { token, network } from '../../properties';
import { WatcherInterface } from '../WatcherInterface';
import * as monitor from 'node-docker-monitor';

export class DockerWatcher extends WatcherInterface{

  constructor() {
    super();
  }

  /**
 * Schaut ob es sich um eine absolute url handelt.(Startet mit http(s)://)
 * Wenn nicht sucht sie die ip des netzwerkes raus.
 * (Relative URL z.B.: :{port}{suburl}(:3001/graphql))
 */
  updateUrl = (url:string, sockData:any) :string => {
    if (url.startsWith('http')) {
      return url;
    }
    return 'http://' + sockData.NetworkSettings.Networks[network()].IPAddress + url;

  }

  onContainerUp = (container: any) => {
    if (container.Labels[clientLabels.TOKEN] === token()) {
      const url = container.Labels[clientLabels.URL];
      const namespace :string = container.Labels[clientLabels.NAMESPACE];
      const deploymentName = container.Id;
      this.deleteEndpoint(namespace, deploymentName);

      if (this.endpoints[namespace] === undefined) {
        this.endpoints[namespace] = [];
      }
      this.endpoints[namespace].push({
        namespace,
        url: this.updateUrl(url, container),
        typePrefix: namespace + '_',
        __created: container.Created,
        __imageID: container.Image,
        __deploymentName: deploymentName,
      });
      this.callDataUpdateListener();
    }
  }

  onContainerDown = (container) => {};
  watchEndpoint = () => {
    monitor({
      onMonitorStarted: () => {},
      onMonitorStopped: () => {},
      onContainerUp: this.onContainerUp,
      onContainerDown: this.onContainerDown,
    });
  }
}
