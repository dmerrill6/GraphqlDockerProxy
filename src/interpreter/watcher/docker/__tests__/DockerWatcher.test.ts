import { DockerWatcher } from '../DockerWatcher'

jest.mock('../../../../properties', () => {
  return {
    token: () => '123',
    network: () => 'web',
  }
})
describe('test the DockerWatcher', () => {

  let watcher : DockerWatcher = null
  beforeEach(() => {
    watcher = new DockerWatcher()
  })

  describe('tests onContainerUp', () => {

    it('tests by container with gql label token', () => {
      const exampleContainer = {        NetworkSettings: {
        Networks: {
          web: {
            IPAddress: '127.0.0.1',
          },
        },
      },

        Labels: {
          'gqlProxy.token': '123',
          'gqlProxy.url': ':mock:9000',
          'gqlProxy.namespace': 'a',
        },
      }
      watcher.callDataUpdateListener = jest.fn()
      watcher.onContainerUp(exampleContainer)
      expect(watcher.callDataUpdateListener).toBeCalled()
      expect(watcher.endpoints).toMatchSnapshot()

    })

    it('tests by container with no gql label token', () => {
      const exampleContainer = {
        Labels: {

        },
      }
      watcher.callDataUpdateListener = jest.fn()
      watcher.onContainerUp(exampleContainer)
      expect(watcher.callDataUpdateListener).not.toBeCalled()

    })
  })

  describe('tests updateUrl', () => {
    it('do the job by absolute url', () => {
      const url = 'http://test.de/graphql'
      expect(watcher.updateUrl(url, {})).toEqual(url)

    })

    it('do the job by relative path', () => {
      const url = ':9000/graphql'
      const result = watcher.updateUrl(url, {
        NetworkSettings: {
          Networks: {
            web: {
              IPAddress: '127.0.0.1',
            },
          },
        },
      })

      expect(result).toEqual('http://127.0.0.1:9000/graphql')
    })
  })

})
