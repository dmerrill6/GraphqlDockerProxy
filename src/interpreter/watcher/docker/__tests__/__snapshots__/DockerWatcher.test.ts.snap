// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test the DockerWatcher tests onContainerUp tests by container with gql label token 1`] = `
Object {
  "a": Array [
    Object {
      "__deploymentName": undefined,
      "__imageID": undefined,
      "namespace": "a",
      "typePrefix": "a_",
      "url": "http://127.0.0.1:mock:9000",
    },
  ],
}
`;
